vm/actor/src/builtin/market/mod.rs

```diff
@@ -478,247 +596,361 @@ impl Actor {
-                    // Not yet appeared in proven sector; check for timeout.
-                    if state.sector_start_epoch == EPOCH_UNDEFINED {
+                    let dcid = deal.cid().map_err(|e| {
+                        actor_error!(ErrIllegalState;
+                                    "failed to calculate cid for proposal {}: {}", deal_id, e)
+                    })?;
+
+                    let state = msm.deal_states.as_ref().unwrap().get(deal_id).map_err(
+                        |e| actor_error!(ErrIllegalState; "failed to get deal state: {}", e),
+                    )?;
+
+                    // deal has been published but not activated yet -> terminate it
+                    // as it has timed out
+                    if state.is_none() {
+                        // Not yet appeared in proven sector; check for timeout.
                         assert!(
-                            rt.curr_epoch() >= deal.start_epoch,
-                            "if sector start is not set, we must be in a timed out state"
+                            curr_epoch >= deal.start_epoch,
+                            "if sector start is not set, must be in timed out state"
                         );
 
-                        let slashed = st.process_deal_init_timed_out(
-                            rt.store(),
-                            &mut et,
-                            &mut lt,
-                            id,
-                            &deal,
-                            state,
-                        )?;
-                        amount_slashed += slashed;
-
+                        let slashed = msm.process_deal_init_timed_out(&deal)?;
+                        if !slashed.is_zero() {
+                            amount_slashed += slashed;
+                        }
                         if deal.verified_deal {
                             timed_out_verified_deals.push(deal.clone());
                         }
-                    }
 
-                    let (slash_amount, next_epoch) = st.update_pending_deal_state(
-                        rt.store(),
-                        state,
-                        deal,
-                        id,
-                        &mut et,
-                        &mut lt,
-                        rt.curr_epoch(),
-                    )?;
-                    amount_slashed += slash_amount;
+                        // we should not attempt to delete the DealState because it does NOT exist
+                        msm.deal_proposals
+                            .as_mut()
+                            .unwrap()
+                            .delete(deal_id)
+                            .map_err(
+                                |e| actor_error!(ErrIllegalState; "failed to delete deal: {}", e),
+                            )?;
+                        msm.pending_deals
+                            .as_mut()
+                            .unwrap()
+                            .delete(&dcid.to_bytes())
```

Comment:
For this delete and all the ones before, should we handle the case where the delete returns false indicating that what you're deleting doesnt exist?

Reply:
good catch, actually didn't realize, and just checked, they just error whenever trying to delete something that doesn't exist (weird). Will update now